File: packages/console/internal/deployment/sites.go
package deployment

import (
	"context"
	"net/http"

	"github.com/better-giving/console/internal/cf"
)

// the sites this deployment's donation forms may be loaded on, as this console writes them back.
//
// **the rule is read in front of the press and it is not read here.** what a site may be — the
// scheme, the trailing slash, the wildcard, the count cap — is one module both ends read, and the
// browser applies it to the boxes before anything is posted (packages/console-ui/src/lib/sites.ts).
// a second reading in this binary would be a third statement of it, and the half that drifts is the
// one nothing reports. **the console reading it first is a courtesy to the operator and never a
// replacement for the boundary**: that path is reachable by anything holding a session, so the
// worker parses every list it is sent whatever asked it to — and its refusal arrives here as one of
// the arms below.
//
// **the whole list goes with every press, never one row.** the endpoint stores a list and deletes
// what is not in it inside the same batch, so a row left out of the body is a site removed. that is
// also why a removal can be refused: a site still ticked on a live donation form is one the
// deployment will not drop, and it answers naming every form standing in the way.
//
// **a press answers with the report, so this console never holds a second view of the list.** the
// stored list comes back off the write itself, which is what the boxes are re-seeded from.
//
// every failure is a value: nothing here returns an error, and an error handed up to a handler
// would be a 500 in place of the state that explains it.

// SitesPath is the path on a deployment that stores its site list.
const SitesPath = "/console/sites"

// SitesWriteKind is how one write of the list ended.
type SitesWriteKind string

const (
	// SitesSaved is the list moving, which is the only kind that changed the deployment.
	SitesSaved SitesWriteKind = "saved"
	// SitesRefused is the deployment reading the list against the rule and turning it down.
	SitesRefused SitesWriteKind = "refused"
	// SitesBlocked is a site still listed on a live donation form, so the whole list was kept.
	SitesBlocked SitesWriteKind = "blocked"
	// SitesUnwritten is a deployment that never read one, and Read says which way.
	SitesUnwritten SitesWriteKind = "unwritten"
)

// BlockingForm is one donation form standing in the way of a removal, as the endpoint names it.
type BlockingForm struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

// BlockedSite is one site that could not be removed, and every form still listing it.
type BlockedSite struct {
	Site  string         `json:"site"`
	Forms []BlockingForm `json:"forms"`
}

// SitesWrite is how one write went.
//
// Flat rather than a member per kind, because that is what the wire is: every field is written on
// every answer and each is empty on the kinds that say nothing about it.
type SitesWrite struct {
	Kind SitesWriteKind `json:"kind"`
	// Sites is the list the deployment holds now, read back off the report the write answered with.
	Sites []string `json:"sites"`
	// Message is the deployment's own sentence about a refusal or a removal it would not make, and
	// Fix its second — the only thing on the screen that says how to get out of the state.
	Message string  `json:"message"`
	Fix     *string `json:"fix"`
	// InUse is every site a live donation form still lists, on SitesBlocked alone.
	InUse []BlockedSite `json:"inUse"`
	// Read is which way the deployment never read a list at all, and nil on every other kind.
	Read *NoReport `json:"read"`
}

// SaveSites stores the list, or says why it did not.
//
// The list is posted as it was typed, never as anything here parsed it: the trim, the blank-row
// skip and the dedupe are the deployment's to make over what it stores, and a console that sent its
// own reading would be storing a list nobody typed.
//
// A nil writer is its own answer rather than a request made with no credential, for the reason
// ./org.go states.
func SaveSites(ctx context.Context, post cf.Post, sites []string) SitesWrite {
	if post == nil {
		return unstored(NoReport{Kind: NoSession})
	}
	return readSitesWrite(post(ctx, SitesPath, map[string]any{"sites": sites}))
}

// what the deployment answered a write with, read.
//
// The classification is the report's and is not restated. What is added in front of it is the two
// answers only this write gets — the refusal, and the removal the deployment would not make.
func readSitesWrite(answer cf.Answer) SitesWrite {
	body, _ := answer.Body.(map[string]any)
	if answer.Kind == cf.Answered && answer.Status == http.StatusUnprocessableEntity {
		return SitesWrite{
			Kind:    SitesRefused,
			Sites:   []string{},
			Message: said(body, "This deployment turned the list down and said nothing this console could read."),
			Fix:     text(body["fix"]),
			InUse:   []BlockedSite{},
		}
	}
	if answer.Kind == cf.Answered && answer.Status == http.StatusConflict {
		return SitesWrite{
			Kind:    SitesBlocked,
			Sites:   []string{},
			Message: said(body, "A site on this list is still in use by a donation form, so nothing was saved."),
			Fix:     text(body["fix"]),
			InUse:   blockedSites(body["inUse"]),
		}
	}

	read := readReport(answer)
	if read.Kind != Reported {
		return unstored(read.NoReport)
	}
	return SitesWrite{Kind: SitesSaved, Sites: read.Sites, InUse: []BlockedSite{}}
}

// the forms the endpoint named, read one row at a time.
//
// The member arrived off a network: a row that is not a form is dropped rather than drawn, and the
// sentence beside it is the deployment's own either way.
func blockedSites(value any) []BlockedSite {
	listed, ok := value.([]any)
	if !ok {
		return []BlockedSite{}
	}
	blocked := []BlockedSite{}
	for _, entry := range listed {
		held, mapped := entry.(map[string]any)
		if !mapped {
			continue
		}
		site := text(held["site"])
		if site == nil {
			continue
		}
		rows, _ := held["forms"].([]any)
		forms := []BlockingForm{}
		for _, row := range rows {
			form, mapped := row.(map[string]any)
			if !mapped {
				continue
			}
			name := text(form["name"])
			if name == nil {
				continue
			}
			id, _ := form["id"].(string)
			forms = append(forms, BlockingForm{ID: id, Name: *name})
		}
		blocked = append(blocked, BlockedSite{Site: *site, Forms: forms})
	}
	return blocked
}

// the deployment's own sentence, or this console's where it wrote none readable.
func said(body map[string]any, otherwise string) string {
	if wrote := text(body["message"]); wrote != nil {
		return *wrote
	}
	return otherwise
}

func unstored(read NoReport) SitesWrite {
	return SitesWrite{Kind: SitesUnwritten, Sites: []string{}, InUse: []BlockedSite{}, Read: &read}
}

File: packages/console-ui/src/lib/sites.ts
import { readOriginRows } from '@better-giving/operator/origins';
import { z } from 'zod';
import type { StatedForm } from './use-console-form';

// the site list as the panel holds it: the group the boxes stand in, the name each row submits
// under, the rules the fold marks them by, what the press posts as its intent, and the list the
// boxes are seeded from.
//
// **the boxes and the rules they are read against are one module, and nothing about a credential is
// in it.** the panel draws the rows, the schema below marks them, and the binary posts them to the
// deployment over the session it holds (`packages/console/internal/deployment/sites.go`) — so the
// vocabulary is stated once, on the side a component can import.
//
// **one press, and it costs seconds.** saving the list posts to a deployment that is already
// running and levels the turnstile widget behind it (`packages/console/internal/widget`); nothing is
// republished and no one-way door is gone through.
//
// **adding and dropping a row are not that press.** they are the form's own list intents, applied
// in the browser by conform and posted nowhere (./use-console-form.ts) — so the boxes are the
// operator's to arrange, and only Save sites leaves this machine.
//
// **the rules are not stated here and are `@better-giving/operator/origins`'s.** that module is
// where both surfaces read them from and its header argues why a copy is the thing that must not
// happen: one end reading `acme.org` as `https://acme.org` and the other refusing the list is a
// setup run that dies after doing work nobody can take back. what this file adds is the shape
// conform reads them through — which sentence is keyed to which box.
//
// **the deployment's parse stays the authority.** `/console/sites` is reachable by anything holding
// a console session, so the worker parses every list it is sent whatever asked it to; the fold
// marking a box first is a courtesy to the person typing.
//
// **the one rule stated here rather than read from that module is that a box must hold something.**
// every box a form states must arrive (CLAUDE.md → Bans → Forms), and this is that rule over a row:
// a blank is refused under its own box and means nothing else. it is not a rule about what a site
// may be — the deployment drops a blank row rather than reading one, which is why it is stated on
// this form and not in the module both ends share. what empties the list is every row being
// dropped, which posts no box at all.
//
// **the press is read for whether the boxes differ from the seed they were drawn with**, which is
// the form layer's own reading and no rule of this module's (./use-console-form.ts), and every
// answer it may get is the same one: post the boxes. an operator moving domains means to empty the
// list, the worker stores an empty one, and a console refusing it would be turning down a value the
// worker accepts.

/** the list field every row of the boxes belongs to, and the name each row is indexed under. */
export const SITE_FIELD = 'site';

/**
 * the group the rows are drawn in, which is also the form's own id.
 *
 * every box id on the fold is composed off it by the seam (./use-console-form.ts), so `sites` is
 * the prefix a row's box is found by and the name the group's own message is drawn under.
 */
export const SITES_BOX = 'sites';

/** what the press that stores the list posts, which is the submitting button's own value. */
export const SITES_INTENT = 'sites';

/**
 * one box of the list as the form layer reads it, which is the one place an empty box is a value.
 *
 * conform drops every empty string before the schema sees it (`parseWithZod` in
 * `@conform-to/zod/v4`), so a bare `z.string()` would answer a row nobody has typed into with
 * "expected string, received undefined" — a refusal over the box the fold opens on. `.optional()`
 * is what lets that row through to the rule below, which is where a blank is dropped rather than
 * marked. ./org-form.ts's `identityBox` states the same reading for a box that stands alone.
 */
const siteRow = z.string().optional();

/**
 * the sentence a box holding nothing gets, and it is one word: the box it is drawn under is
 * labelled by the group it stands in, so a sentence naming the field again is the label spelled
 * twice under itself.
 */
const REQUIRED = 'required';

const sitesBoxes = z.object({
	/**
	 * the rows read as one list, because two of the things that can be wrong with them are facts
	 * about the list rather than about a row: how many there are, and a site named in a box above. a
	 * rule mounted on the row could see neither.
	 *
	 * every sentence is keyed to the box it is about, which is what a path under this field does:
	 * conform spells `['site', 1]` as `site[1]`, the second row's own input, and the fold draws it
	 * under that box. the one sentence keyed to the bare field is the cap on how many sites there
	 * may be, which is a fact about no one row — and because it is keyed to a name no control on the
	 * form carries, it is the one message conform's submit-time focus move cannot reach, which is
	 * why the fold draws it at the button rather than over the boxes (./sites-fold.tsx).
	 *
	 * **a blank row is marked here and nowhere else.** `readOriginRows` hands one back unmarked —
	 * a blank is not a site, so the module both surfaces read has nothing to say about it — and this
	 * form refuses it, so the row is marked before that reading is taken.
	 */
	[SITE_FIELD]: z.array(siteRow).superRefine((rows, ctx) => {
		// the empty string a blank box stood for, put back for the reader: conform drops every empty
		// string before the schema sees it, and this is the shape `readOriginRows` reads.
		const lines = rows.map((row) => row ?? '');
		lines.forEach((line, at) => {
			if (line.trim().length > 0) return;
			ctx.addIssue({ code: 'custom', message: REQUIRED, path: [at] });
		});

		const read = readOriginRows(lines);
		if (read.problem !== null) {
			ctx.addIssue({ code: 'custom', message: read.problem });
			return;
		}
		read.rows.forEach((problem, at) => {
			if (problem === null) return;
			ctx.addIssue({ code: 'custom', message: problem, path: [at] });
		});
	})
});

/** the sites fold's form: what its `<Form>` is called, and the rules its press runs first. */
export const SITES_FORM: StatedForm<typeof sitesBoxes> = {
	id: SITES_BOX,
	schema: sitesBoxes
};

/**
 * the boxes the fold opens on, which is the list the deployment holds and nothing else.
 *
 * stated here rather than in the fold because the button's reading is taken against it: what counts
 * as an edit is a box differing from the one it was drawn with, and a seed spelled twice is a fold
 * whose button and whose boxes disagree about what the operator has done.
 *
 * **a deployment holding no site opens on no box at all**, because the group is never empty: the
 * locked row is drawn above these and the fold reads as one row and an Add. a blank box seeded over
 * it would be a row nobody added, offering a Remove for nothing and refusing itself the moment the
 * press was reached. the locked row is the fold's and is in no list here: it posts nothing
 * (./sites-fold.tsx).
 */
export function siteSeed(stored: readonly string[]): string[] {
	return [...stored];
}

/** the name one row's box submits under, which is the list field indexed by the row's position. */
const ROW_NAME = new RegExp(`^${SITE_FIELD}\\[\\d+\\]$`);

/**
 * whether a control on this form is one of the list's rows, by the name it submits under.
 *
 * the one spelling of that name, read by the fold off its own boxes and here off what they posted.
 * the locked row is in neither: it carries no name at all
 * (`fixed` in `@better-giving/operator/components/forms/RepeatingRows`).
 */
export const isSiteRow = (name: string): boolean => ROW_NAME.test(name);

/**
 * what the boxes hold, in the order they are on screen.
 *
 * nothing is trimmed, nothing is dropped and nothing is refused. an empty row is kept because it is
 * a box an operator is still typing into, and what a row may hold is answered at the box rather
 * than here.
 *
 * the rows are a list field, so each box submits under its own position — `site[0]`, `site[1]` —
 * rather than every one of them under a bare `site`. the order is the boxes' own: a form's entries
 * arrive in the order the controls stand in the document.
 */
export function siteEdits(posted: FormData): string[] {
	return [...posted]
		.filter(([name]) => isSiteRow(name))
		.map(([, row]) => (typeof row === 'string' ? row : ''));
}
